numpy/typing/__init__.py

```diff
@@ -131,6 +131,14 @@
 else:
     def final(f): return f
 
+if not TYPE_CHECKING:
+    __all__ = ["ArrayLike", "DTypeLike", "NBitBase"]
+else:
+    # Ensure that all objects within this module are accessible while
+    # static type checking. This includes private ones, as we need them
+    # for internal use.
+    __all__: List[str]
```

Comment:
What does this do?

Reply:
If `__all__` is defined then mypy will only consider the objects mentioned therein to be importable. 
Obviously this is a problem, as the `numpy.typing` contains a number of private variables that we _don't_ want in `__all__`.

As a workaround, `__all__` is now annotated as a generic list-of-strings while type checking, thus circumventing the issue above. In this manner we can still access our private variables and mypy knows that there is an `__all__` object within the module.

Reply:
What I find curious is the use of `:`.

Reply:
I suppose that is an annotation here. Perhaps make the explanation more explicit.

Reply:
Done as of https://github.com/numpy/numpy/pull/17844/commits/670d94ffba5798559bff3538dba39ab01f050248.